metricq/exceptions.py

```diff
@@ -0,0 +1,116 @@
+#     * Neither the name of metricq nor the names of its contributors
+#       may be used to endorse or promote products derived from this software
+#       without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""
+This module defines custom Exception classes for MetricQ.
+In general, they are used for errors in the interaction with other agents.
+Contrary, misuse of the API generally yields specific built-in errors, i.e., TypeError, ValueError, and, KeyError.
+AssertionError / assert is only used for checking invariants within the library itself.
+"""
+
+
+class RemoteError(Exception):
+    """The remote has replied with an error."""
+
+
+class RPCError(RemoteError):
+    """An RPC returned an error code.
+
+    This could be an issue with the input or on the other side.
+    """
+
+    pass
+
+
+class HistoryError(RemoteError):
+    """The database replied with an error to a history request.
+
+    This could either be an issue with the request or of the database.
+    """
+
+    pass
+
+
+class MessageError(Exception):
+    """Something is semantically wrong within a received message."""
+
+    pass
+
+
+class InvalidHistoryResponse(MessageError):
+    """A response to a history request could not be decoded."""
+
+    def __init__(self, extra_message):
+        super().__init__(
+            (
+                "Failed to parse inconsistent history response message: {}."
+                + "This is probably not the callers fault, but caused by the db."
+            ).format(extra_message)
+        )
+
+
+class NonMonotonicTimestamps(MessageError):
+    """Timestamps in a history response are not strictly monotonic."""
+
+
+class AgentStopped(Exception):
+    """The agent was stopped unexpectedly.
+
+    There is probably something wrong with the network.
+    """
+
+    pass
+
+
+class ReceivedSignal(AgentStopped):
+    """The agent was stopped by a specific signal."""
+
+    def __init__(self, signal, *args):
+        self.signal = signal
+        super().__init__(f"Received signal {signal} while running Agent", *args)
+
+
+class ConnectFailed(AgentStopped):
+    """The connection attempt of the agent failed."""
+
+    pass
+
+
+class ReconnectTimeout(AgentStopped):
```

Comment:
Shouldn't there be a `ConnectTimeout` as well?

Reply:
There's `ConnectFailed` for that.